### ISSUE TITLE
Creating tcp socket beside a opc ua server

### ISSUE THREAD

**Issue by muurc** [open62541/open62541#6366]
## Description
Currently, I am working on a project in which my server is going to get TCP packages from another TCP server and subsequently displays the data on an information model. So far, I've tried to implement an extra TCP socket with the arpa/inet header and maintain the management in another pthread.  However, after connecting with the TCP server, both event loops seem to shut down. It can create the tcp socket, yet there is no connection to the server. It then seems to jump back to the OPCUA event loop and stop the socket there. Edit: It now connects and communicates with the server to some extend, but still shuts down.

Is there even a way to implement this TCP client plugin alongside the OPCUA server? If yes, should I use the arpa/inet structure, or is there maybe a way to use the OPCUA library for this? And perhaps if this is not makeable, is there any other way you could recommend me doing to receive TCP packages?

I appreciate every suggestion.

## Background Information / Reproduction Steps
Creating a OPCUA server, then implementing a arpa socket and running the thread in a pthread.

Used CMake options:

```bash
cmake 
-DBUILD_SHARED_LIBS=ON 
-DCMAKE_BUILD_TYPE=RelWithDebInfo 
-DUA_NAMESPACE_ZERO=FULL 
-UA_ENABLE_MULTITHREAD=ON ..
```

## Checklist

Please provide the following information:

 - [x ] open62541 Version (release number or git tag): 1.4.0
 - [ ] Other OPC UA SDKs used (client or server): 
 - [x ] Operating system: Ubuntu 20.04.6
 - [ ]
![Screenshot from 2024-03-27 08-46-12](https://github.com/open62541/open62541/assets/165135492/3cbd049b-6096-4231-8e51-fc5e50793dd6)
 Logs (with `UA_LOGLEVEL` set as low as necessary) attached
 - [ ] Wireshark network dump attached
 - [ ] Self-contained code example attached
 - [ ] Critical issue

**Comment by jpfr** [comment]
Yes. The EventLoop was made exactly for that.
https://github.com/open62541/open62541/blob/master/include/open62541/plugin/eventloop.h

Every server and client configuration contains an EventLoop.
You can interact with that and run your socket in parallel to the normal server operations.

You can reopen this issue if there are further questions.

**Comment by muurc** [comment]
Hello Mister Pfrommer,

I hope you dont mind me asking, but I was wondering if it is possible to
send "normal" tcp messages from an open62541 client to a server, and what feature would
I need for implementing a message handler on the server side.

I would appreciate a suggestion from your side.

Sincerely,

Marc Futschik
________________________________
Von: Julius Pfrommer ***@***.***>
Gesendet: Dienstag, 2. April 2024 13:49
An: open62541/open62541 ***@***.***>
Cc: Marc Futschik ***@***.***>; Author ***@***.***>
Betreff: Re: [open62541/open62541] Creating tcp socket beside a opc ua server (Issue #6366)


Yes. The EventLoop was made exactly for that.
https://github.com/open62541/open62541/blob/master/include/open62541/plugin/eventloop.h

Every server and client configuration contains an EventLoop.
You can interact with that and run your socket in parallel to the normal server operations.

You can reopen this issue if there are further questions.

**Comment by jpfr** [comment]
Only TCP would not be OPC UA.
But if you need it, check the examples here: https://github.com/open62541/open62541/blob/master/tests/check_eventloop_tcp.c
You can do the same with the EventLoop that is driving every open62541 client and server.
Basically you open a connection with a callback method that gets triggered when an event happens (message arrives, connection closes, etc.).

**Comment by muurc** [comment]
Thank you very much again. I have been trying to find the right solution for a whole day now.
So your showing me the correct approach certainly helped me.

Have a nice day, and weekend!

Marc Futschik
________________________________
Von: Julius Pfrommer ***@***.***>
Gesendet: Donnerstag, 4. April 2024 22:23
An: open62541/open62541 ***@***.***>
Cc: Marc Futschik ***@***.***>; Author ***@***.***>
Betreff: Re: [open62541/open62541] Creating tcp socket beside a opc ua server (Issue #6366)


Only TCP would not be OPC UA.
But if you need it, check the examples here: https://github.com/open62541/open62541/blob/master/tests/check_eventloop_tcp.c
You can do the same with the EventLoop that is driving every open62541 client and server.
Basically you open a connection with a callback method that gets triggered when an event happens (message arrives, connection closes, etc.).
